lua/nvim-treesitter/textobj.lua

```diff
@@ -0,0 +1,158 @@
+    end_col = #vim.fn.getline('$')
+  end
+
+  vim.fn.setpos(".", { buf, start_row + 1, start_col + 1, 0 })
+  vim.fn.nvim_exec("normal v", false)
+
+  -- Convert exclusive end position to inclusive
+  if end_col == 0 then
+    vim.fn.setpos(".", { buf, end_row, -1, 0 })
+  else
+    vim.fn.setpos(".", { buf, end_row + 1, end_col, 0 })
+  end
+end
+
+local function is_in_node_range(node, line, col)
+  local start_line, start_col, end_line, end_col = node:range()
+  if line >= start_line and line <= end_line then
+    if line == start_line and line == end_line then
+      return col >= start_col and col < end_col
+    elseif line == start_line then
+      return col >= start_col
+    elseif line == end_line then
+      return col < end_col
+    else
+      return true
+    end
+  else
+    return false
+  end
+end
+
+local function node_length(node)
+  local _, _, start_byte = node:start()
+  local _, _, end_byte = node:end_()
+  return end_byte - start_byte
+end
+
+------------ should go to ts_utils
+
+function M.select_textobj(query_string)
+  local bufnr = vim.api.nvim_get_current_buf()
+  local ft = api.nvim_buf_get_option(bufnr, "ft")
+  if not ft then return end
+  local lang = parsers.ft_to_lang(ft)
+
+  local row, col = unpack(vim.api.nvim_win_get_cursor(0))
+  row = row - 1
+
+  local matches = {}
+
+
+  if string.match(query_string, '^@.*') then
+    local query = vim.split(string.sub(query_string, 2), '%.')
+    table.insert(query, 'node')
+    for _, m in pairs(locals.get_locals(bufnr)) do
+      local insert = m
+      for _, q in pairs(query) do
+        if type(insert) == 'table' then
+          insert = insert[q]
+        end
+      end
+      if insert then
+        table.insert(matches, insert)
+      end
+    end
+  else
+    local parser = parsers.get_parser(bufnr, lang)
+    local root = parser:parse():root()
+    local start_row, _, end_row, _ = root:range()
+
+    local nested = {}
+    local query = ts.parse_query(lang, query_string)
+    for prepared_match in queries.iter_prepared_matches(query, root, bufnr, start_row, end_row) do
+      table.insert(nested, prepared_match)
+    end
+    for _, m in pairs(nested) do
+      for _, n in pairs(m) do
+        if n.node then
+          table.insert(matches, n.node)
+        end
+      end
+    end
+  end
+
+  local match_length
+  local smallest_range
+
+  for _, m in pairs(matches) do
+    if is_in_node_range(m, row, col) then
+      local length = node_length(m)
+      if not match_length or length < match_length then
+        smallest_range = m
+        match_length = length
+      end
+    end
+  end
+
+  if smallest_range then
+    update_selection(bufnr, smallest_range)
+  end
+end
+
+function M.attach(bufnr)
+    local buf = bufnr or api.nvim_get_current_buf()
+    local config = configs.get_module("textobj")
+
+    local lang = parsers.ft_to_lang(api.nvim_buf_get_option(bufnr, "ft"))
+    for mapping, queries in pairs(config.keymaps) do
```

Comment:
maybe just check `config.keymaps` is a table, if its not return, so then you dont have to check if the query exists because you'd know you would have something in your keymap :) You could do the same for the detach function.
This is just a suggestion, i like to have less control flow in loops as possible as it makes the code easier to understand (to me at least).